File: sdk/src/resources/telemetry.ts
import type { HttpClient } from '../http.js';
import type {
  TelemetryChannel,
  TelemetryReading,
  TelemetryNotification,
  UpsertTelemetryChannelParams,
  ListTelemetryReadingsParams,
  TelemetryReadingInput,
  IngestTelemetryReadingsResult,
  ListTelemetryNotificationsParams,
  PaginatedResponse,
  SingleResponse,
} from '../types.js';
import { paginate } from '../paginator.js';

export class TelemetryResource {
  constructor(private readonly http: HttpClient) {}

  async listChannels(params: { assetId?: string } = {}): Promise<TelemetryChannel[]> {
    const response = await this.http.get<{ success: boolean; data: TelemetryChannel[] }>(
      '/telemetry/channels',
      params as Record<string, unknown>
    );
    return response.data;
  }

  /** Create or update a telemetry channel by (assetId, key). Requires write scope. */
  async upsertChannel(params: UpsertTelemetryChannelParams): Promise<TelemetryChannel> {
    const response = await this.http.post<SingleResponse<TelemetryChannel>>('/telemetry/channels', params);
    return response.data;
  }

  async listReadings(params: ListTelemetryReadingsParams = {}): Promise<PaginatedResponse<TelemetryReading>> {
    return this.http.get<PaginatedResponse<TelemetryReading>>(
      '/telemetry/readings',
      params as Record<string, unknown>
    );
  }

  listAllReadings(params: Omit<ListTelemetryReadingsParams, 'page'> = {}): AsyncGenerator<TelemetryReading> {
    return paginate((p) => this.listReadings(p), params);
  }

  /**
   * Ingest up to 1000 readings in a single call.
   * CRIT breaches auto-escalate the asset to NFPA 70B Condition 2.
   * Requires write scope. Use idempotencyKey to make retries safe.
   */
  async ingestReadings(
    readings: TelemetryReadingInput[],
    idempotencyKey?: string
  ): Promise<IngestTelemetryReadingsResult> {
    const response = await this.http.post<SingleResponse<IngestTelemetryReadingsResult>>(
      '/telemetry/readings',
      { readings },
      idempotencyKey
    );
    return response.data;
  }

  async listNotifications(params: ListTelemetryNotificationsParams = {}): Promise<{
    data: TelemetryNotification[];
    count: number;
  }> {
    const response = await this.http.get<{ success: boolean; data: TelemetryNotification[]; count: number }>(
      '/telemetry/notifications',
      params as Record<string, unknown>
    );
    return { data: response.data, count: response.count };
  }

  /** Acknowledge a WARN/CRIT notification. Requires write scope. */
  async acknowledgeNotification(id: string): Promise<TelemetryNotification> {
    const response = await this.http.post<SingleResponse<TelemetryNotification>>(
      `/telemetry/notifications/${id}/acknowledge`
    );
    return response.data;
  }
}

File: sdk/python/servicecycle/resources/telemetry.py
"""Mirrors sdk/src/resources/telemetry.ts."""

from __future__ import annotations

from typing import Any, Dict, Iterator, List, Optional

from .._util import camel_params
from ..http import HttpClient
from ..paginator import paginate


class TelemetryResource:
    def __init__(self, http: HttpClient):
        self._http = http

    def list_channels(self, **params: Any) -> List[Dict[str, Any]]:
        """Kwargs: asset_id, page, limit(<=500, default 200)."""
        return self._http.get("/telemetry/channels", camel_params(params))["data"]

    def upsert_channel(self, **params: Any) -> Dict[str, Any]:
        """Create or update a channel by (asset_id, key). Requires write scope.

        Kwargs: asset_id, key (required), label, unit, warn_high, crit_high,
        warn_low, crit_low, enabled.
        """
        return self._http.post("/telemetry/channels", camel_params(params))["data"]

    def list_readings(self, **params: Any) -> Dict[str, Any]:
        """Kwargs: asset_id, channel, since, page, limit(<=500)."""
        return self._http.get("/telemetry/readings", camel_params(params))

    def list_all_readings(self, **params: Any) -> Iterator[Dict[str, Any]]:
        return paginate(lambda p: self._http.get("/telemetry/readings", camel_params(p)), params)

    def ingest_readings(self, readings: List[Dict[str, Any]], idempotency_key: Optional[str] = None) -> Dict[str, Any]:
        """Ingest 1-1000 readings in a single batch call. CRIT breaches
        auto-escalate the asset's governing condition to at least C2.
        Requires write scope. Pass idempotency_key for safe retries.

        Each reading dict: {asset_id, channel, value, unit?, recorded_at?,
        source?, external_id?}.

        Returns: {accepted, breaches, duplicates, total, results: [...]}.
        """
        body = {"readings": [camel_params(r) for r in readings]}
        return self._http.post("/telemetry/readings", body, idempotency_key)["data"]

    def list_notifications(self, **params: Any) -> Dict[str, Any]:
        """Kwargs: status('open'|'all', default 'open'), asset_id.
        Returns (no pagination envelope -- hard-capped at 200 server-side):
        {data: [...], count}.
        """
        raw = self._http.get("/telemetry/notifications", camel_params(params))
        return {"data": raw["data"], "count": raw.get("count")}

    def acknowledge_notification(self, notification_id: str) -> Dict[str, Any]:
        """Acknowledge a WARN/CRIT notification. Requires write scope."""
        return self._http.post(f"/telemetry/notifications/{notification_id}/acknowledge")["data"]
